File: Data Structure/Array or Matrix/Hard/LET_no850_Rectangle Area II.py
class Solution:
    def rectangleArea(self, rectangles: List[List[int]]) -> int:
        xs, ys = [], []
        for x1, y1, x2, y2 in rectangles:
            xs.extend([x1, x2])
            ys.extend([y1, y2])
            
        xs = sorted(set(xs))
        ys = sorted(set(ys))
        x_i = {x : i for i, x in enumerate(xs)}
        i_x = {i : x for i, x in enumerate(xs)}
        y_j = {y : j for j, y in enumerate(ys)}
        j_y = {j : y for j, y in enumerate(ys)}
        
        board = [[0] * len(ys) for _ in range(len(xs))]
        for x1, y1, x2, y2 in rectangles:
            i1, i2, j1, j2 = x_i[x1], x_i[x2], y_j[y1], y_j[y2]
            for i in range(i1, i2):
                for j in range(j1, j2):
                    board[i][j] = 1
                    
        res, mod = 0, 10 ** 9 + 7
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == 1:
                    res += (i_x[i + 1] - i_x[i]) * (j_y[j + 1] - j_y[j])
                    res %= mod
        
        return res

File: Data Structure/Array or Matrix/Hard/LET_no850_Rectangle Area II.cpp
class Solution {
public:
    int rectangleArea(vector<vector<int>>& rectangles) {
        set<int> xs, ys;
        for (auto rec : rectangles)
            xs.insert(rec[0]), ys.insert(rec[1]), xs.insert(rec[2]), ys.insert(rec[3]);
        
        unordered_map<int, int> x_i, y_j, i_x, j_y;
        int i_count = 0, j_count = 0;
        for (auto x = xs.begin(); x != xs.end(); x++) {
            x_i[*x] = i_count;
            i_x[i_count] = *x;
            i_count++;
        }
        for (auto y = ys.begin(); y != ys.end(); y++) {
            y_j[*y] = j_count;
            j_y[j_count] = *y;
            j_count++;
        }
        
        vector<vector<int>> board(i_count, vector<int>(j_count, 0));
        for (auto rec : rectangles) {
            int i1 = x_i[rec[0]], j1 = y_j[rec[1]], i2 = x_i[rec[2]], j2 = y_j[rec[3]];
            for (int i = i1; i < i2; i++) {
                for (int j = j1; j < j2; j++)
                    board[i][j] = 1;
            }
        }
        
        long long res = 0;
        int mod = (int) (1e9 + 7);
        for (int i = 0; i < i_count; i++) {
            for (int j = 0; j < j_count; j++) {
                if (board[i][j]) {
                    res += ((long long)(i_x[i + 1] - i_x[i])) * (j_y[j + 1] - j_y[j]);
                    res %= mod;
                }
            }
        }
        
        return res;
    }
};